### ISSUE TITLE
exorelay-go

### ISSUE THREAD

**Issue by kevgo** [Originate/exosphere-sdk#124]
- use `json.RawMessage` to represent JSON data in Go?
- decode it into map[string]interface{}?
- code generation?

**Comment by arekkas** [comment]
where and why do you need to represent the JSON data? what would a receiver method signature look like?

**Comment by arekkas** [comment]
in http, a handlerfunc looks like:

```go
func myHandler(w http.ResponseWriter, r *http.Request)
```

similarily, I think that a request callback handler in exo would look like:

*exorelay-go/handler.go:*
```go
package exo

type Reader interface/struct ...
type Writer interface/struct ...

type ExorelayHandler func(r Reader, w Writer)
```

*someproject/main.go:*
```go
import "github.com/Originate/exorelay-go"

func main() {
    server := exo.NewServer()
    server.HandleFunc("message_id", myHandler)
    server.Listen()
}

func myHandler(r exo.Reader, w exo.Writer) {
    var myData fooStruct
    if err := json.NewDecoder(r.Body).Decode(&myData); err != nil {
        // ...err
    }
}
```

**Comment by kevgo** [comment]
> where and why do you need to represent the JSON data? what would a receiver method signature look like?

This would work similar to https://github.com/originate/exorelay-js. Here is how the JavaScript version is used:

- your code creates an ExoRelay instance:

  ```ls
  exoRelay = new ExoRelay exocomIp: '127.0.0.1', exocomPort: 3001, serviceName: 'test'
  ```

- send out a message with some payload:

  ```ls
  exoRelay.send 'hello', name: 'world'
  ```

  This sends the string "{name: "hello", payload: { name: "world" }" over a ZMQ socket to 127.0.0.1:3001.

- register a handler for incoming messages:

  ```ls
  exoRelay.registerHandler 'hello', (name) -> console.log "hello #{name}!"
  ```

  This calls the given callback function each time a message with name "hello" is received

**Comment by arekkas** [comment]
ok, it depends if you want json to be abstracted away or not. from what i see and what we talked about (replacing json with protobuf in the future) I think you want it to be abstracted away. this is kind of possible with go, but en-/decoders sometimes need additional information to decode successfully (think of ORMs like Doctrine/Hybernate with annotations). In go, this is true to:

```go
type myData struct {
 Foo string `json:"foo_field"`
 Bar string `json:"bar"`
}
```

The JSON encoder can only read exported fields, which will then be uppercase `"Bar": ...`. That's why you need the json annotations. Go is probably not the only language where this is required. I haven't worked with protobuf yet, there might be similar "limitations".

Seeing your code example, I would do:

```go
srv := exo.Relay{
  IP: "12.0.0.1", // this should be ipv6 able
  Port: 3001,
  Name: "test"
}

srv.Handle("message", myHandler)

func myHandler(r exo.Reader, w exo.Writer) {
    var myData fooStruct
    if err := r.Read(&myData); err != nil {
        // ...err
    }

    if err := w.Write(&myData); err != nil {
        // ...err
    }
}
```

**Comment by arekkas** [comment]
There's a second approach which uses prototypes and interfaces, but i'm 90% confident the go community will kill you for that.

```go
srv := exo.Relay{
  IP: "12.0.0.1", // this should be ipv6 able
  Port: 3001,
  Name: "test"
}

srv.Handle("message", myHandler, fooStruct{})

func myHandler(r exo.Reader, w exo.Writer) {
    var myData interface{} = r.Body

   // type assertion neccessary here
   // myTypedData, ok := myData.(*fooStruct)
   // if !ok {
   // ....


    if err := w.Write(&myData); err != nil {
        // ...err
    }
}
```

**Comment by arekkas** [comment]
I think the code in https://github.com/Originate/exosphere-sdk/issues/124#issuecomment-240665926 is pretty close to the js version and having a dedicate `Read` function is quite idiomatic to go. I would probably roll with that or a varaition of it.

I understand now better what exoRealy is doing and how everything is working. Honest to heart, I was weirded out by the syntax in the examples which made it harder for me to understand. Maybe break this down in a similar fashion in the docs? or use regular es5/es6 as everyone can read that?

**Comment by arekkas** [comment]
Please take a look at https://golang.org/pkg/net/http/#HandleFunc and https://golang.org/pkg/net/http/#Handle . One is to have an immediate callback and the other is for a reference. You might want that, too.

**Comment by kevgo** [comment]
Alright, thanks for all the support on this! We try to get basic a Go version up as soon as we find somebody to work on this. 

> I was weirded out by the syntax in the examples which made it harder for me to understand

Are you referring to the LiveScript examples? Yes, they should be converted to ES6.

**Comment by arekkas** [comment]
Yup, LiveScript - makes sense once you know it, but it's another syntax :)
